### PR TITLE
Unionize with interfaces

### PR DESCRIPTION
**Pull request by KnorpelSenf**
Motivation: https://github.com/telegraf/telegraf/pull/1115#issuecomment-679030254

About half the work cherry-picked from #3.

Recommend to review commit-by-commit.

### CODE REVIEW

**Comment by KnorpelSenf** [comment]
@wojpawlik should `Chat.MultiUserGetChat.permissions` really be kept optional at this point? The docs are not clear about this IMO

https://github.com/KnorpelSenf/typegram/pull/5/files#diff-d96ec7ae5bb4a0907779f25ed03acb33R179

File: types.d.ts [hunk 1]
```diff
@@ -1,23 +1,31 @@
...
     author_signature?: String;
-    chat: { type: "channel" };
-    from?: never;
+    from: never;
   }
+  /** Internal type holding properties that updates outside of channels share. */
   interface NonChannel {
-    author_signature?: never;
+    chat: { type: Exclude<Chat["type"], Update.Channel["chat"]["type"]> };
+    author_signature: never;
```
Comment (wojpawlik) [inline types.d.ts]: It's impossible to create a value of type `never`, so this `interface` is impossible to instantiate!
```suggestion
    author_signature?: never;
```

File: types.d.ts [hunk 2]
```diff
@@ -1,23 +1,31 @@
...
     from: User;
   }
+  /** Internal type holding properties that updates about new messages share. */
+  interface New {
+    edit_date: never;
```
Comment (wojpawlik) [inline types.d.ts]: It's impossible to create a value of type `never`, so this `interface` is impossible to instantiate!
```suggestion
    edit_date?: never;
```

File: types.d.ts [hunk 3]
```diff
@@ -1,23 +1,31 @@
 import { Boolean, Float, Integer, String, True } from "./alias";
 
 declare namespace Update {
+  /** Internal type holding properties that updates in channels share. */
   interface Channel {
+    chat: { type: Chat.ChannelChat["type"] };
     author_signature?: String;
-    chat: { type: "channel" };
-    from?: never;
+    from: never;
```
Comment (wojpawlik) [inline types.d.ts]: It's impossible to create a value of type `never`, so this `interface` is impossible to instantiate!
```suggestion
    from?: never;
```

File: types.d.ts [hunk 4]
```diff
@@ -1,23 +1,31 @@
...
-    forward_from_message_id?: never;
-    forward_signature?: never;
-    forward_sender_name?: never;
-    forward_date?: never;
+    forward_from: never;
+    forward_from_chat: never;
+    forward_from_message_id: never;
+    forward_signature: never;
+    forward_sender_name: never;
+    forward_date: never;
```
Comment (wojpawlik) [inline types.d.ts]: It's impossible to create a value of type `never`, so this `interface` is impossible to instantiate!
```suggestion
    forward_from?: never;
    forward_from_chat?: never;
    forward_from_message_id?: never;
    forward_signature?: never;
    forward_sender_name?: never;
    forward_date?: never;
```

File: types.d.ts [hunk 5]
```diff
@@ -1,33 +1,97 @@
 import { Boolean, Float, Integer, String, True } from "./alias";
 
+declare namespace Update {
+  /** Internal type holding properties that updates in channels share. */
+  interface Channel {
+    chat: { type: Chat.ChannelChat["type"] };
```
Comment (wojpawlik) [inline types.d.ts]:
```suggestion
    chat: Chat.ChannelChat;
```

File: types.d.ts [hunk 6]
```diff
@@ -1,23 +1,31 @@
...
     author_signature?: String;
-    chat: { type: "channel" };
-    from?: never;
+    from: never;
   }
+  /** Internal type holding properties that updates outside of channels share. */
   interface NonChannel {
-    author_signature?: never;
+    chat: { type: Exclude<Chat["type"], Update.Channel["chat"]["type"]> };
```
Comment (wojpawlik) [inline types.d.ts]:
```suggestion
    chat: Exclude<Chat, Chat.ChannelChat>;
```

**Review by wojpawlik — CHANGES_REQUESTED** [review]
1. > should `Chat.MultiUserGetChat.permissions` really be kept optional at this point?

   Unsure. Leaving it optional is the safe choice.

2. After dec0043, this is no longer expressed in types, is it intentional?
   ```
   For backward compatibility, when this field is set, the document / location field will also be set
   ```

3. `KeyboardButton` and `InlineKeyboardButton` could also be unionized here.

4. I suggest squash-merging this, or cleaning up with interactive rebase after we're done improving the PR.

5. Tip: you can add suggestions to a batch to accept multiple suggestions in a single commit.

6. Heads up: I have one more idea how to improve `Message`, I'll open a PR after this is merged.

File: types.d.ts [hunk 7]
```diff
@@ -74,142 +138,305 @@ export interface UserFromGetMe extends User {
...
-  sticker_set_name?: String;
-  /** True, if the bot can change the group sticker set. Returned only in getChat. */
-  can_set_sticker_set?: Boolean;
+export type ChatFromGetChat =
+  | Chat.PrivateGetChat
+  | Chat.GroupGetChat
+  | Chat.SupergroupGetChat
+  | Chat.ChannelGetChat;
+
+declare namespace Message {
```
Comment (wojpawlik) [inline types.d.ts]: These types will be very useful for https://github.com/telegraf/telegraf/pull/1116
```suggestion
export namespace Message {
```

**Comment by KnorpelSenf** [comment]
> 2. After [dec0043](https://github.com/KnorpelSenf/typegram/commit/dec0043ace99b6d207c173d6b4fac4ee1dcbc074), this is no longer expressed in types, is it intentional?
>    ```
>    For backward compatibility, when this field is set, the document / location field will also be set
>    ```

Yes it is, I'm extending the respective type instead. That's why every `venue` also has `location` set and every `animation` also has `document` set.

> 3. `KeyboardButton` and `InlineKeyboardButton` could also be unionized here.

I agree, I might have a look at it tomorrow (unless you take care of it in the meantime)

> 4. I suggest squash-merging this, or cleaning up with interactive rebase after we're done improving the PR.

I agree, I vote for rebasing interactively in this case.

> 5. Tip: you can add suggestions to a batch to accept multiple suggestions in a single commit.

Thanks! Didn't check that out yet, will do

> 6. Heads up: I have one more idea how to improve `Message`, I'll open a PR after this is merged.

Looking forward to it!

File: types.d.ts (hunk 7 above)
Comment (KnorpelSenf) [reply types.d.ts]: We could generally just export the namespaces. There's not really any “internal logic” in this lib that needs to be hidden. If someone wants to dig into how that types are built and leverage that structure, I don't see why we should prevent that by not exporting something.

**Review by wojpawlik — APPROVED** [review]
I'm not permitted to push here, so to avoid blocking, I'll take care of `(Inline)?KeyboardButton` in the next PR. Feel free to rebase interactively, forcepush and merge.

**Pull request merged**